File: platforms/android/lib/src/main/java/com/shopify/checkoutkit/Telemetry.kt
package com.shopify.checkoutkit

import android.webkit.WebViewClient.ERROR_CONNECT
import android.webkit.WebViewClient.ERROR_HOST_LOOKUP
import android.webkit.WebViewClient.ERROR_IO
import android.webkit.WebViewClient.ERROR_TIMEOUT
import com.shopify.checkoutkit.telemetry.CheckoutKitTelemetry
import com.shopify.checkoutkit.telemetry.TelemetryErrorCode
import com.shopify.checkoutkit.telemetry.TelemetryErrorMetric
import com.shopify.checkoutkit.telemetry.TelemetryNavigationDurationMetric
import com.shopify.checkoutkit.telemetry.TelemetryNavigationRetryMetric
import com.shopify.checkoutkit.telemetry.TelemetryNavigationRetryReason
import com.shopify.checkoutkit.telemetry.TelemetryPlatform
import com.shopify.checkoutkit.telemetry.TelemetryProtocolDecodeErrorMetric

internal interface CheckoutTelemetryRecording {
    fun recordError(metric: TelemetryErrorMetric)
    fun recordProtocolDecodeError(metric: TelemetryProtocolDecodeErrorMetric)
    fun recordNavigationRetry(metric: TelemetryNavigationRetryMetric)
    fun recordNavigationDuration(metric: TelemetryNavigationDurationMetric)
}

private interface CheckoutTelemetryClient : CheckoutTelemetryRecording {
    fun shutdown()
}

private class DefaultCheckoutTelemetryRecorder : CheckoutTelemetryClient {
    private val telemetry = CheckoutKitTelemetry(
        sdkVersion = BuildConfig.SDK_VERSION,
        platform = if (ShopifyCheckoutKit.configuration.platform is Platform.ReactNative) {
            TelemetryPlatform.ReactNativeAndroid
        } else {
            TelemetryPlatform.Android
        },
    ).also { it.start() }

    override fun recordError(metric: TelemetryErrorMetric) = telemetry.recordError(metric)

    override fun recordProtocolDecodeError(metric: TelemetryProtocolDecodeErrorMetric) =
        telemetry.recordProtocolDecodeError(metric)

    override fun recordNavigationRetry(metric: TelemetryNavigationRetryMetric) =
        telemetry.recordNavigationRetry(metric)

    override fun recordNavigationDuration(metric: TelemetryNavigationDurationMetric) =
        telemetry.recordNavigationDuration(metric)

    override fun shutdown() = telemetry.shutdown(discardPending = true)
}

private object NoOpCheckoutTelemetryRecorder : CheckoutTelemetryRecording {
    override fun recordError(metric: TelemetryErrorMetric) = Unit

    override fun recordProtocolDecodeError(metric: TelemetryProtocolDecodeErrorMetric) = Unit

    override fun recordNavigationRetry(metric: TelemetryNavigationRetryMetric) = Unit

    override fun recordNavigationDuration(metric: TelemetryNavigationDurationMetric) = Unit
}

internal object CheckoutTelemetry {
    private val lock = Any()

    @Volatile
    private var configuredRecorder: CheckoutTelemetryClient? = null

    @Volatile
    private var recorderOverride: CheckoutTelemetryRecording? = null

    val recorder: CheckoutTelemetryRecording
        get() {
            if (!ShopifyCheckoutKit.configuration.telemetry.enabled) {
                return NoOpCheckoutTelemetryRecorder
            }
            recorderOverride?.let { return it }
            configuredRecorder?.let { return it }
            return synchronized(lock) {
                // Re-check under the lock so a concurrent disable() cannot race a
                // recorder creation that would keep exporting after opt-out.
                if (!ShopifyCheckoutKit.configuration.telemetry.enabled) {
                    return NoOpCheckoutTelemetryRecorder
                }
                recorderOverride
                    ?: configuredRecorder
                    ?: runCatching { DefaultCheckoutTelemetryRecorder() }
                        .getOrNull()
                        ?.also { configuredRecorder = it }
                    ?: NoOpCheckoutTelemetryRecorder
            }
        }

    fun disable() {
        val recorder = synchronized(lock) {
            configuredRecorder.also { configuredRecorder = null }
        }
        recorder?.shutdown()
    }

    fun overrideRecorderForTesting(recorder: CheckoutTelemetryRecording?) {
        recorderOverride = recorder
    }

    fun errorCode(errorCode: Int?): TelemetryErrorCode = when (errorCode) {
        ERROR_TIMEOUT -> TelemetryErrorCode.Timeout
        ERROR_CONNECT -> TelemetryErrorCode.CannotConnect
        ERROR_HOST_LOOKUP -> TelemetryErrorCode.Dns
        ERROR_IO -> TelemetryErrorCode.ConnectionLost
        else -> TelemetryErrorCode.Unknown
    }

    fun retryReason(errorCode: Int?): TelemetryNavigationRetryReason = when (errorCode(errorCode)) {
        TelemetryErrorCode.Timeout -> TelemetryNavigationRetryReason.Timeout
        TelemetryErrorCode.ConnectionLost -> TelemetryNavigationRetryReason.ConnectionLost
        TelemetryErrorCode.CannotConnect -> TelemetryNavigationRetryReason.CannotConnect
        TelemetryErrorCode.Dns -> TelemetryNavigationRetryReason.Dns
        else -> TelemetryNavigationRetryReason.Unknown
    }
}

File: platforms/swift/Sources/ShopifyCheckoutKit/Telemetry.swift
#if !COCOAPODS
    import CheckoutKitTelemetry
#endif
import Foundation

protocol CheckoutTelemetryRecording: Sendable {
    func recordError(_ metric: TelemetryErrorMetric)
    func recordProtocolDecodeError(_ metric: TelemetryProtocolDecodeErrorMetric)
    func recordNavigationRetry(_ metric: TelemetryNavigationRetryMetric)
    func recordNavigationDuration(_ metric: TelemetryNavigationDurationMetric)
}

private protocol CheckoutTelemetryClient: CheckoutTelemetryRecording {
    func start()
    func shutdown(discardPending: Bool) async -> Bool
}

extension CheckoutKitTelemetry: CheckoutTelemetryClient {}

private struct NoOpCheckoutTelemetryRecorder: CheckoutTelemetryRecording {
    func recordError(_: TelemetryErrorMetric) {}
    func recordProtocolDecodeError(_: TelemetryProtocolDecodeErrorMetric) {}
    func recordNavigationRetry(_: TelemetryNavigationRetryMetric) {}
    func recordNavigationDuration(_: TelemetryNavigationDurationMetric) {}
}

private struct CheckoutTelemetryState {
    var client: (any CheckoutTelemetryClient)?
    var recorderOverride: (any CheckoutTelemetryRecording)?
}

/// Stamps an entry point's product onto every measurement before forwarding to
/// the shared client, so one client instance serves every entry point.
private struct ProductScopedRecorder: CheckoutTelemetryRecording {
    let product: TelemetryProduct
    let base: any CheckoutTelemetryRecording

    func recordError(_ metric: TelemetryErrorMetric) {
        base.recordError(
            .init(
                category: metric.category,
                stage: metric.stage,
                code: metric.code,
                retryable: metric.retryable,
                isRetry: metric.isRetry,
                product: product
            )
        )
    }

    func recordProtocolDecodeError(_ metric: TelemetryProtocolDecodeErrorMetric) {
        base.recordProtocolDecodeError(
            .init(method: metric.method, failureType: metric.failureType, product: product)
        )
    }

    func recordNavigationRetry(_ metric: TelemetryNavigationRetryMetric) {
        base.recordNavigationRetry(
            .init(reason: metric.reason, result: metric.result, product: product)
        )
    }

    func recordNavigationDuration(_ metric: TelemetryNavigationDurationMetric) {
        base.recordNavigationDuration(
            .init(
                milliseconds: metric.milliseconds,
                result: metric.result,
                preloaded: metric.preloaded,
                product: product
            )
        )
    }
}

private let noOpCheckoutTelemetryRecorder = NoOpCheckoutTelemetryRecorder()
private let lockedCheckoutTelemetry = LockedValue(CheckoutTelemetryState())

enum CheckoutTelemetry {
    static var recorder: any CheckoutTelemetryRecording {
        recorder(for: nil)
    }

    static func recorder(for entryPoint: MetaData.EntryPoint?) -> any CheckoutTelemetryRecording {
        guard ShopifyCheckoutKit.configuration.telemetry.enabled else {
            return noOpCheckoutTelemetryRecorder
        }

        var base: (any CheckoutTelemetryRecording)?
        lockedCheckoutTelemetry.update { state in
            if let recorderOverride = state.recorderOverride {
                base = recorderOverride
                return
            }
            if let client = state.client {
                base = client
                return
            }

            // Re-check under the lock so a concurrent disable() cannot race a
            // client creation that would keep exporting after opt-out.
            guard ShopifyCheckoutKit.configuration.telemetry.enabled else {
                return
            }

            let client = CheckoutKitTelemetry(
                configuration: .init(
                    sdkVersion: MetaData.version,
                    platform: telemetryPlatform()
                )
            )
            client.start()
            state.client = client
            base = client
        }
        guard let base else { return noOpCheckoutTelemetryRecorder }
        return ProductScopedRecorder(
            product: entryPoint == .acceleratedCheckouts ? .acceleratedCheckouts : .checkoutKit,
            base: base
        )
    }

    static func disable() {
        var client: (any CheckoutTelemetryClient)?
        lockedCheckoutTelemetry.update { state in
            client = state.client
            state.client = nil
        }
        if let client {
            Task { _ = await client.shutdown(discardPending: true) }
        }
    }

    static func overrideRecorderForTesting(_ recorder: (any CheckoutTelemetryRecording)?) {
        lockedCheckoutTelemetry.update { state in
            state.recorderOverride = recorder
        }
    }

    private static func telemetryPlatform() -> TelemetryPlatform {
        ShopifyCheckoutKit.configuration.platform?.identifier == "ReactNative"
            ? .reactNativeSwift
            : .swift
    }

    static func errorCode(for error: NSError) -> TelemetryErrorCode {
        guard error.domain == NSURLErrorDomain else { return .unknown }
        switch error.code {
        case NSURLErrorCancelled: return .cancelled
        case NSURLErrorTimedOut: return .timeout
        case NSURLErrorNetworkConnectionLost: return .connectionLost
        case NSURLErrorCannotConnectToHost: return .cannotConnect
        case NSURLErrorDNSLookupFailed: return .dns
        default: return .unknown
        }
    }

    fileprivate static func retryReason(for error: NSError) -> TelemetryNavigationRetryReason {
        switch errorCode(for: error) {
        case .timeout: return .timeout
        case .connectionLost: return .connectionLost
        case .cannotConnect: return .cannotConnect
        case .dns: return .dns
        default: return .unknown
        }
    }
}

extension TelemetryNavigationRetryMetric {
    /// Derives the bounded retry reason from the navigation error, so call
    /// sites record retries without mapping errors to reasons themselves.
    init(error: NSError, result: TelemetryNavigationRetryResult) {
        self.init(reason: CheckoutTelemetry.retryReason(for: error), result: result)
    }
}
